File: sdk-java/src/test/java/io/superdurable/dex/integ/StepCancellationTest.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ;

import io.superdurable.dex.FlowResult;
import io.superdurable.dex.StepExecutionId;
import io.superdurable.dex.testing.DexDevTestEnvironment;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.Duration;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

@Tag("dex-dev")
public final class StepCancellationTest {
    private static final Duration START_TIMEOUT = Duration.ofSeconds(10);
    private static final Duration CANCELLATION_TIMEOUT = Duration.ofSeconds(8);
    private static final Duration FLOW_TIMEOUT = Duration.ofSeconds(30);

    @TempDir
    Path cacheDirectory;

    @Test
    void testHeartbeatCancelsRegularExecuteAndInterruptsHandler() throws Exception {
        assertCooperativeCancellation(
                StepCancellationWorkflow.Scenario.HEARTBEAT_EXECUTE,
                CANCELLATION_TIMEOUT);
    }

    @Test
    void testHeartbeatCancelsRegularWaitForAndInterruptsHandler() throws Exception {
        assertCooperativeCancellation(
                StepCancellationWorkflow.Scenario.HEARTBEAT_WAIT_FOR,
                CANCELLATION_TIMEOUT);
    }

    @Test
    void testLocalActivityCancellationInterruptsWithoutFallback() throws Exception {
        final StepCancellationWorkflow workflow = new StepCancellationWorkflow(
                StepCancellationWorkflow.Scenario.LOCAL_EXECUTE);
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                workflow)) {
            final String flowId = start(environment, workflow, "cancel-local");
            assertTrue(workflow.awaitBlockingHandlerStarted(START_TIMEOUT));
            environment.client().waitForStepCompletion(
                    flowId,
                    StepExecutionId.of(workflow.canceledStepType()),
                    CANCELLATION_TIMEOUT);
            assertTrue(workflow.awaitCancellation(CANCELLATION_TIMEOUT));
            assertCompleted(environment, flowId, StepCancellationWorkflow.Scenario.LOCAL_EXECUTE);
            assertTrue(workflow.wasHandlerInterrupted());
            assertTrue(workflow.didContextReportCancellation());
            assertEquals(1, workflow.blockingExecuteInvocations());
            assertFalse(workflow.wasRecoveryRun());
            assertNull(environment.client().getAttribute(flowId, workflow.lateWrite));
        }
    }

    @Test
    void testTimerCancelsAfterLocalTimeoutFallback() throws Exception {
        final StepCancellationWorkflow workflow = new StepCancellationWorkflow(
                StepCancellationWorkflow.Scenario.LOCAL_TIMEOUT_FALLBACK);
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                workflow)) {
            final String flowId = start(environment, workflow, "cancel-local-fallback");
            assertTrue(workflow.awaitBlockingHandlerStarted(START_TIMEOUT));
            environment.client().waitForStepCompletion(
                    flowId,
                    StepExecutionId.of(workflow.canceledStepType()),
                    FLOW_TIMEOUT);
            assertCompleted(
                    environment,
                    flowId,
                    StepCancellationWorkflow.Scenario.LOCAL_TIMEOUT_FALLBACK);
            assertEquals(1, workflow.blockingExecuteInvocations());
            assertTrue(workflow.wasHandlerInterrupted());
            assertTrue(workflow.didContextReportCancellation());
            assertFalse(workflow.wasRecoveryRun());
            assertNull(environment.client().getAttribute(flowId, workflow.lateWrite));
        }
    }

    @Test
    void testCancellationWithoutHeartbeatContinuesBeforeLateHandlerReturn() throws Exception {
        final StepCancellationWorkflow workflow = new StepCancellationWorkflow(
                StepCancellationWorkflow.Scenario.NO_HEARTBEAT);
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                workflow)) {
            final String flowId = start(environment, workflow, "cancel-no-heartbeat");
            assertTrue(workflow.awaitBlockingHandlerStarted(START_TIMEOUT));
            environment.client().waitForStepCompletion(
                    flowId,
                    StepExecutionId.of(workflow.canceledStepType()),
                    CANCELLATION_TIMEOUT);
            assertCompleted(environment, flowId, StepCancellationWorkflow.Scenario.NO_HEARTBEAT);
            assertFalse(workflow.hasLateHandlerReturned());
            assertFalse(workflow.wasHandlerInterrupted());
            assertTrue(workflow.awaitLateHandlerReturn(CANCELLATION_TIMEOUT));
            assertEquals(1, workflow.blockingExecuteInvocations());
            assertFalse(workflow.wasRecoveryRun());
            assertNull(environment.client().getAttribute(flowId, workflow.lateWrite));
        }
    }

    @Test
    void testGlobalSelectorCancelsMatchingExecutionsFromBothParents() throws Exception {
        final StepCancellationWorkflow workflow = new StepCancellationWorkflow(
                StepCancellationWorkflow.Scenario.GLOBAL_SELECTOR);
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                workflow)) {
            final String flowId = start(environment, workflow, "cancel-global");
            assertTrue(workflow.awaitSelectorWaits(START_TIMEOUT));
            environment.client().publish(flowId, workflow.selectorWinnerRelease, (Void) null);
            assertCompleted(environment, flowId, StepCancellationWorkflow.Scenario.GLOBAL_SELECTOR);
            assertFalse(workflow.wasFirstSelectorExecuted());
            assertFalse(workflow.wasSecondSelectorExecuted());
        }
    }

    @Test
    void testSiblingSelectorKeepsMatchingExecutionFromOtherParent() throws Exception {
        final StepCancellationWorkflow workflow = new StepCancellationWorkflow(
                StepCancellationWorkflow.Scenario.SIBLING_SELECTOR);
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                workflow)) {
            final String flowId = start(environment, workflow, "cancel-sibling");
            assertTrue(workflow.awaitSelectorWaits(START_TIMEOUT));
            environment.client().publish(flowId, workflow.selectorWinnerRelease, (Void) null);
            environment.client().waitForStepCompletion(
                    flowId,
                    StepExecutionId.of(workflow.selectorWinnerStepType()),
                    CANCELLATION_TIMEOUT);
            environment.client().publish(flowId, workflow.selectorWaitingRelease, (Void) null);
            assertTrue(workflow.awaitSecondSelectorExecution(CANCELLATION_TIMEOUT));
            environment.client().publish(flowId, workflow.selectorFinalRelease, (Void) null);
            assertCompleted(environment, flowId, StepCancellationWorkflow.Scenario.SIBLING_SELECTOR);
            assertFalse(workflow.wasFirstSelectorExecuted());
            assertTrue(workflow.wasSecondSelectorExecuted());
        }
    }

    private void assertCooperativeCancellation(
            final StepCancellationWorkflow.Scenario scenario,
            final Duration cancellationTimeout) throws Exception {
        final StepCancellationWorkflow workflow = new StepCancellationWorkflow(scenario);
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                workflow)) {
            final String flowId = start(environment, workflow, "cancel-heartbeat");
            assertTrue(workflow.awaitBlockingHandlerStarted(START_TIMEOUT));
            environment.client().waitForStepCompletion(
                    flowId,
                    StepExecutionId.of(workflow.canceledStepType()),
                    cancellationTimeout);
            assertTrue(workflow.awaitCancellation(cancellationTimeout));
            assertCompleted(environment, flowId, scenario);
            assertTrue(workflow.wasHandlerInterrupted());
            assertTrue(workflow.didContextReportCancellation());
            assertFalse(workflow.wasRecoveryRun());
            assertNull(environment.client().getAttribute(flowId, workflow.lateWrite));
        }
    }

    private static String start(
            final DexDevTestEnvironment environment,
            final StepCancellationWorkflow workflow,
            final String prefix) {
        final String flowId = prefix + "-" + UUID.randomUUID();
        environment.client().startFlow(workflow, flowId, null);
        return flowId;
    }

    private static void assertCompleted(
            final DexDevTestEnvironment environment,
            final String flowId,
            final StepCancellationWorkflow.Scenario scenario) {
        final FlowResult result = environment.client().waitForFlow(flowId, FLOW_TIMEOUT);
        assertEquals(scenario.name(), result.getSingleOutput(String.class));
    }
}

File: sdk-rust/crates/dex-sdk/tests/integ/step_cancellation_test.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::sync::Arc;
use std::sync::atomic::Ordering;
use std::time::Duration;

use dex_sdk::{Registry, StepExecutionId};

use crate::step_cancellation_workflow::{
    CancellationBlockingExecute, CancellationBlockingWaitFor, CancellationScenario,
    CancellationState, LATE_WRITE, StepCancellationWorkflow,
};
use crate::support::{DexDevTestEnvironment, flow_id};

#[test]
#[ignore = "requires dexcli dev"]
fn test_step_cancellation() {
    for scenario in CancellationScenario::ALL {
        run_scenario(scenario);
    }
}

fn run_scenario(scenario: CancellationScenario) {
    let state = CancellationState::new(scenario);
    let workflow = StepCancellationWorkflow::new(Arc::clone(&state));
    let environment = DexDevTestEnvironment::start(Registry::new().register(workflow));
    let client_workflow = StepCancellationWorkflow::new(Arc::clone(&state));
    let flow_id = flow_id(&format!("rust-cancellation-{}", scenario.name()));
    environment
        .client
        .start_flow(&client_workflow, &flow_id, scenario.name().to_string())
        .expect("start cancellation Flow");

    if !matches!(
        scenario,
        CancellationScenario::GlobalSelector | CancellationScenario::SiblingSelector
    ) {
        assert!(state.blocking_started.wait(Duration::from_secs(10)));
        if scenario == CancellationScenario::HeartbeatWaitFor {
            environment
                .client
                .wait_for_step_completion(
                    &flow_id,
                    StepExecutionId::of(&CancellationBlockingWaitFor(Arc::clone(&state))),
                    Duration::from_secs(30),
                )
                .expect("wait for canceled WaitFor Step");
        } else {
            environment
                .client
                .wait_for_step_completion(
                    &flow_id,
                    StepExecutionId::of(&CancellationBlockingExecute {
                        state: Arc::clone(&state),
                    }),
                    Duration::from_secs(30),
                )
                .expect("wait for canceled Execute Step");
        }
    }

    assert_eq!(
        scenario.name(),
        environment
            .client
            .wait_for_flow_with_timeout(&flow_id, Duration::from_secs(30))
            .and_then(|result| result.single_output::<String>())
            .expect("complete cancellation Flow")
    );

    match scenario {
        CancellationScenario::GlobalSelector => {
            assert!(!state.first_selector_executed.load(Ordering::SeqCst));
            assert!(!state.second_selector_executed.load(Ordering::SeqCst));
            return;
        }
        CancellationScenario::SiblingSelector => {
            assert!(!state.first_selector_executed.load(Ordering::SeqCst));
            assert!(state.second_selector_executed.load(Ordering::SeqCst));
            return;
        }
        CancellationScenario::NoHeartbeat => {
            assert!(!state.handler_canceled.load(Ordering::SeqCst));
            assert!(!state.late_handler_returned.is_set());
            assert!(state.late_handler_returned.wait(Duration::from_secs(8)));
        }
        _ => {
            assert!(state.cancellation_observed.wait(Duration::from_secs(8)));
            assert!(state.handler_canceled.load(Ordering::SeqCst));
            assert!(state.context_reported_cancellation.load(Ordering::SeqCst));
        }
    }
    let expected_invocations = if matches!(
        scenario,
        CancellationScenario::LocalExecute | CancellationScenario::LocalTimeoutFallback
    ) {
        2
    } else {
        1
    };
    assert_eq!(
        expected_invocations,
        state.blocking_invocations.load(Ordering::SeqCst),
        "{}",
        scenario.name()
    );
    assert!(!state.recovery_ran.load(Ordering::SeqCst));
    assert_eq!(
        None,
        environment
            .client
            .get_attribute::<String>(&flow_id, &LATE_WRITE)
            .expect("read late Attribute")
    );
}
